### ISSUE TITLE
WYSIWIG Editors poorly designed and implemented.

### ISSUE THREAD

**Issue by GarfieldS** [SocialEngine/phpv4-issues#278]
## Wysiwig editors are a mess.

I have been facing a lot of grief over this and it is very understandable. Here is a link to the issue with images. http://bit.ly/1YOMiyr 

I will be as clear as possible. 

Regular users are able to upload images using the blog editor with many advanced options but Admins are limited to a rich text editor with:
- no upload options
- no video embed options
- no source code view
- no  wide area editing space 
- no blockquote

The HTML editor expect one to create their html content somewhere else or hand code it on the spot with no preview using a text area that is too small to effectively edit this content without scrolling frequently and losing your space if the content is even of a normal size blog post.

To add images in the html editor one needs to:
- Switch to the File and image media manager to upload images
- Go back to the rich text editor after losing their place or have the forethought to upload images first
- Past the code they hopefully remembered to copy. (multiple images makes this more complex)

Again, all this is done in a small textarea box that makes it difficult to create content without scrolling down for an even normal sized content. 

The wysiwig edtors are not uniform on the site. I don't understand why the blog editor used by users is more full featured than the rich text editor used by the admin of the site. Anyone who uses wordpress would be LMAO. 

I am sorry to give this tone but when contacting customer support they considered this a "bug" as if they were unaware of the design of the program by default. 

This impacts the everyday use of my client I promoted this solution to and leaves me feeling embarrassed that something that should be so simple continues to frustrate them every time they open up the layout editor. The feeling that this needs to even be explained is infuriating because anyone who uses the program would know this is an issue on first use.

**Comment by pulsatemedia** [comment]
I don't have much to add but a +1 when it comes to the TinyMCE\WYSIWYG in SE. In it's current state, it  is huge user experience fail and I often find myself feeling the same way when discussing this with my users: Embarrassed.

I haven't observed different behavior\features for admins vs users though. That is interesting.

**Comment by GarfieldS** [comment]
Unless I am mistaken, pulsetemedia are you telling me that in your "admin dashboard layout editor" you can: 

* upload images in your rich text editor?
* view the source code in your rich text editor?
* add videos in your rich text editor?
* add block quotes in your rich text editor?
* work in a window that is as large as the window in the blog text editor? 

All of this a regular member can do using the blog wysiwig editor however if an admin dares to do the same in the "layout editor" they are reminded they have less capability than a normal user.

**Comment by pulsatemedia** [comment]
OK I think I understand better now. You are talking specifically about the rich html blocks that can be added in layout editor.

I've had to modify the tinymce.php files in various locations to tweak the appearance\functions of the editor in various modules. You might be able to do same for this feature if you can find the tinymce.php (or similar)

Otherwise, I think this would be relatively straightforward for SE to resolve.

**Comment by GarfieldS** [comment]
I am many others who share the same need would appreciate being able to make the modifications you suggested if we could find these files. You are already a testament to that what was provided to you by default was so lacking for you that you took matters into your own hands. I would love to see a tip for this or a change in the next upgrade that would not require one to start messing with hard coded files that might be overwritten on an update.

> Otherwise, I think this would be relatively straightforward for SE to resolve.

One would hope. However,

**Comment by pulsatemedia** [comment]
I wonder if integrating TinyMCE "enterprise" could be an option (technically and financially) in SE core.....  

http://www.tinymce.com/enterprise/enterprise.php

I know the formatting inconsistency drives my community members nuts, so this might be a win for many folks "Microsoft® Word copy-and-paste with images: unrivalled technology that cleans your pasted word content and matches styles to your preference."


At the very least it's probably worth migrating the current version that comes with SE (4.0.8 2013-10-10) to current 4.2.6 (8/15/15). Theres a plethora is issues resolved in that time: http://www.tinymce.com/develop/changelog/?type=tinymce

For fun I might place the latest tinymce.js on one of my dev sites and see if it works, or breaks things.

So in short, Socialengine: Please, please, please look into improving the WYSIWYG editor that comes with SE core. The exisiting one really takes away from what is otherwise a pretty decent user experience in this product.

**Comment by troydbertrand** [comment]
Socialengine falls short in many areas and I do support some of the points you made here but I think some of your criticism is a bit misguided.

The "HTML Block" wasn't meant to be used as an editor or validator or converter etc. It's simply meant to be use as a block in which raw HTML can be added. The block assumes that you already know that the end product of the raw HTML will be. As a user of WordPress I can tell you that even the most advanced page builders (eg. Visual Composer) use HTML blocks and they're pretty much bare bones. 

I could extend this comment with a lot of examples and analogies but I really don't think it's necessary. 

Regards
TDB

**Comment by GarfieldS** [comment]
I understand what you are saying and as an HTML block it does what it should. I have no criticism about it as a component to copy and paste pre-made HTML. I am not asking for it's removal and my feedback is mostly focused wysiwig editors as written in the title of the post. 

The HTML Block is a poor substitute for adding creating and editing content when there is not a code view for a wysiwig editor. This was just my point for anyone who would recommend it as an alternative to the wysiwig.  You know as well as I do that a lot of customers on Ning use Ning's "text box" to not only copy and paste code like you could with the HTML block but also edit & create content. However, even in HTML view Ning's text box is more useful than the plain HTML block there is a menu to assist in creating content. 

My criticism is for anyone who would suggest that limitations in the rich text editor don't require a better wysiwig editor but could be easily resolved by using the HTML block which some people have been forced to use with much frustration and scrolling.

However, let's not our difference about the HTML block take attention away from where the focus needs to be put on and that is the wysiwig editors.

**Comment by shivani2910** [comment]
Hi @GarfieldS,

I'm so sorry that it took us a while to respond to this thread. This somehow got piled-up with many new issues. 

I agree that there are some toolbar options missing in the TinyMCE editor provided in the Rich Text HTML widget. I've forwarded this issue to our development team and will be working on improving this feature in the next version. Thanks for your feedback.

For now, some of the options mentioned by you in the 1st message to this issue (**video embed options, source code view and blockquote**) can be added to this particular TinyMCE editor by doing a few minor code level changes:

1) Please open file ```/application/modules/Core/Form/Admin/Widget/RichText.php```

2) Search below code at line no. 20:

```
'toolbar1' => array('bold', 'italic', 'underline', 'strikethrough', '|', 'fontselect', 'fontsizeselect', 'forecolor', '|', 'bullist', 'numlist', 'outdent', 'indent', '|', 'alignleft', 'aligncenter', 'alignright', 'alignjustify', '|', 'link', 'image'),
	'plugins' => array('textcolor', 'link', 'image'),
```
Now, replace above code with below one:

```
'toolbar1' => array('bold', 'italic', 'underline', 'strikethrough', '|', 'fontselect', 'fontsizeselect', 'forecolor', '|', 'bullist', 'numlist', 'outdent', 'indent', '|', 'alignleft', 'aligncenter', 'alignright', 'alignjustify', '|', 'link', 'image',  'code', 'media', 'jbimages', 'fullscreen', 'preview', 'blockquote'),
	'plugins' => array('textcolor', 'link', 'image', 'fullscreen', 'media', 'preview', 'code'),
```

3) Now open file ```application/modules/Core/externals/styles/admin/main.css```

4) And add below code just at the end of above file (this is for **wide area editing space in the editor**):

```
#global_page_core-admin-content-widget #global_content_simple > div {width:670px;}
```
Hope that helps. I'll forward the suggestions to our development team and will be working on these enhancements in future.

Thanks,
Shivani

**Comment by GarfieldS** [comment]
Thank you for the reply and your suggested solution until the development team is able to address this in the next version. It is not a minor oversight in my opinion but seems to have been treated as such for such a long time it borders on effecting the credibility of the product. It's even impacted my ability to even promote Social Engine. Imagine this conversation:

**Me**: I've got this really great CMS. In fact, I recommend you nominate if for Critics Choice award at CMS Critic.

**Potential Customer** - Awesome!, How is their WYSIWIG Editor? 

**Me** - Well, it's really good. 

**Potential Customer** - Can you upload image? 

**Me** - No

**Potential Customer** - Can you view and edit the HTML? 

**Me** - No

**Potential Customer** - Can you embed You Tube Videos using it? 

**Me** - No

**Potential Customer** - WTF?!? and you are asking me to nominate this as a Critics Choice for a CMS. Don't you know the CM in CMS stands for Content Management? Did you contact them about this.?  
 
**Me** - Yeah, I wrote an article back in May [Click here] (http://community.onsocialengine.com/post/3294977/give-admins-the-same-editor-members-use-for-blog-posts). However, they just got on Github and I posted the issue there about 2 weeks ago. 

**Potential Customer** - 2 weeks ago?

**Me** - Yeah but they just got back in touch with me today and suggest I edit the hard code to resolve this. Somehow my post got piled-up with many new issues. 

**Potential Customer** - Many new issues? Dude you should have your head examined. I can pick up a CMS on CodeCanyon for the cost of a pair of sneakers and you best believe I can upload some some pictures on it  and you coming at me on the last day of nominations telling me to recommend this to CMS Critic. I don't even want to ask you how much you paid for it. Man, that is weak. Wake and smell the coffee. And you waiting 4 -5 months after bringing this issue up to get it addressed. Are you freaking kidding me? 

Get the picture. I want to fight for you but I can't even hold up under basic questioning. 

Also I recommend that you classify as a bug whatever has you (or anyone on here responsible for managing this Github) overlook a post that has 7 comments on it for 2 weeks and get if fixed. I know you may be busy but this is was my first and ever post on Github ever and it's gonna be like this maybe you need to get a better application. 

I'm patient to a a point and then I start to wonder if you guys are even interested. 

For everyone else, Today is the last day to nominate Social Engine for the CMS Critic award [Click here] (http://awards.cmscritic.com/nominate/) to nominate Social Engine before it's too late

Good luck

**Comment by pulsatemedia** [comment]
I like socialengine, but there's no way I could possibly vote for it as a CMS. Social network, yes, of course. CMS? Nope.

**Comment by GarfieldS** [comment]
@pulsatemedia Thank you for sharing your thoughts on this. Congrats on the blinking cursor!

**Comment by GarfieldS** [comment]
> Hope that helps. I'll forward the suggestions to our development team and will be working on these 
>  enhancements in future.

> Thanks,
> Shivani

I made the changes you suggested. I had a feeling that the upload option might require more customization. From my limited understanding I would imagine that since this plugin has never been installed in the dashboard some adjustment might be need in terms of the destination of the upload. 

Is there any guidance you can provide. Did this work correctly for you on your test installation. When I do my test the upload process times out. I've included the image below so you can see the error. 

![uploading](http://api.ning.com:80/files/T75NQOT1hHQ127fiYLvDBa3ZUDSZds52WmTWOj1IoExj8JFSjeelUPst2zQcbBAXwPgCLuA324OTuHF9W4H4lJG68Wx11uJs/uploadinprogress.jpg)

**Comment by shivani2910** [comment]
Hi @GarfieldS, regarding your query "I had a feeling that the upload option might require more customization..." : Yes, you're right. That's the reason that I said that only these options will be achieved: **"video embed options, source code view and blockquote"**. It seems that apart from the code level changes I've sent to you, you've additionaly inserted **jbimages** to the **plugins** array. Since I did not included this in the code I provided, option to upload images did not appear on my test installation. While it was mistakenly included in toolbar array, but that would not make any difference until it's added to plugins array as well.

So basically, adding the uploading feature to this TinyMCE editor would require more work to be done and I've conveyed this feature to our development team and they'll be implementing it in next version.

Thanks,
Shivani

**Comment by GarfieldS** [comment]
>That's the reason that I said that only these options will be achieved: "video embed options, source 
> code view and blockquote". It seems that apart from the code level changes I've sent to you, you've 
> additionally inserted jbimages to the plugins array. 

You are absolutely right. I did take initiative in adding the upload option there. I understand that this will require more work and you will add this in the update. I will either wait for it or see if I can figure it out on my own. 

Thank you.

**Comment by auntialias** [comment]
Good discussion. Agree about the issues with the WSYWYG editor being an issue. @shivani2910, in the code you referred to, above, is it possible to change settings to instead make the editor allow for semantic markup (H2, H3, H4, etc) versus text size?

**Comment by Elshara** [comment]
Seems to me you could take an html reference like this one http://www.w3schools.com/tags/ and implement most if not all tags available there into a custom html editor then add a wyciwyg overlay on top of it so when you edit something, no matter if it is a list, a table, a paragraph or a place to paste or upload content, it is always displayed from within the same space. I'm tired of using pop up windows for everything and tinymce loves to bring those up. Some users come from Ning and as you may know, everything is edited inline which is extremely useful because if you need to change something, you don't have to worry about column widths, you just have to drop it down and close it when you are finished. I think adding an editor with full html enabled would be much easier than trying to customize a wyciwyg editor which is only programmed to allow certain html tags inside it using a plugin, reference, css and other codes. HTML is ultimately what you get when all of this is combined, it should be a lot easier to add a uniform editor to make this happen, particularly when what most users want to do is add some style and paragraphs to their posts. Yes you can make lists, headings, tables and all sorts of fancy stuff, but when it comes down to it, html should be previewed as wyciwyg but entered like blocks on a page so it all falls into place depending on your needs. Then you can just move the blocks around by pointing your cursor and selecting them, which would in turn give you a custom width and height if you want to change such a thing. If you're for example, adding an embedded video or pasting in a link, what better way to do that than to simply drag it where you want it to appear and resize it based on how you shape it to select its space. This could easily be arranged using an html wyciwyg block manager.

**Comment by GarfieldS** [comment]
I found a solution that allows you to upload pictures directly from the text editor without having to use the Files and Media Manager. It is very simple to implement (takes less than 5 minutes to install)  If you'd like more info contact me. http://goo.gl/forms/tqrjcrzacW

**Comment by troydbertrand** [comment]
I love the idea of having some kind of workaround in the interim but I hope this will not sidetrack the much needed improvements in store for the current WYSIWYG editor. 
TDB

**Comment by GarfieldS** [comment]
It should not in the least. This has nothing to do with Social Engine and would not sidetrack their intended and already declared plans to provide improvements to their current WYSIWYG editor any more than if a Social Engine user editing their own css would derail their intended roadmap for 4.9.0.

**Comment by troydbertrand** [comment]
I know it has nothing to do with SocialEngine. But they'll be less inclined to invest resources to develop or enhance a feature if there's a third party solution available. SocialEngine is usually quick to point out availability of requested features in the form of third party plugin when there's a petition for a new feature or feature enhancement.  "Declarations of Plans" doesn't necessarily mean written in stone. Sidetracking the feature enhancement doesn't mean abandonment. I'm just concerned that it may be bumped down couple notches from the priority list if there's a solution available.

TDB

**Comment by GarfieldS** [comment]
Thank you for clarifying because I was truly perplexed. It seemed like this was what you are saying but the thought of it truly terrified me like a trailer for the next James wan horror movie perhaps like The Conjuring 2. 

However, it is clear that you are describing a situation where a company who just won best social network and has an editor by  with missing features that should be standard that who has stated that they "will be working on improving this feature in the next version."  would abandon it because a freelance developer came up with a solution they couldn't come up with withing a reasonable time?

I don't think Social Engine is that kind of company and I would find that really hard to believe. The addition of this feature is not going to break Social Engine or the internet. It didn't even break me and I can be very fragile sometimes.  I know what you are saying but I don't believe they would not have this resolved in the next version no matter what 3rd party developer solution we to come before the next version.

**Comment by Elshara** [comment]
I wouldn't be surprised if comet chat came out with a text editor that is better than the tinymce social engine one. Which by the way is not consistent in all editing areas. Tinymce plays absolute havic when you have got another editor on the page, it dominates it and prevents any other, be it activity feed status updates, comet chat or the chat box from functioning properly.
I think wyciwyg in general is a bad idea. There's so many versions of the same html alternative, I'd rather just code straight in html with short cut buttons to select text then wrap it as a click and drag format. This would save so much time than having to scroll up and down to get to where you want to format it by sections. I wish working with a blank template was possible. Especially when it comes to embed codes, it just isn't possible in social engine. You can't embed anything in any kind of public post, it sucks. I'd much rather create a blog about something than have to use a verified external link to do it as it should remain something separate.
To be honest there's nothing I do like about creating a post in social engine. the non consistent editor situation is not a deal breaker for me, but I've had more fun with notepad.
To be honest, I'd like an editor that was responsive. What you post is what you get, auto posted code becomes html without having to use a separate editor. Auto copy paste fonts post without issue. Text automatically resizes as you scroll it. The editor can have multiple version if itself on a page without incident and is consistent. When you hit the return key, it forms a new paragraph if the period or other punctuation is detected as the last character, otherwise puts a new line break in. For quotes and images and other basic functions, including tables, it's just html code you fill in what you want it to appear as by selecting the text and wrapping the tags around it. Something simple like that would be perfect.

**Comment by DonnaScriptTechs** [comment]
This is fixed for 4.9 when it is released.

**Comment by Elshara** [comment]
Awesome.
